File: Advent of Code/2022/day 6/solution.ts
import * as fs from "fs";


function read_file_lines(file_name: string = "./input.txt"): string[] {
    const input = fs.readFileSync(file_name, "utf-8");
    const input_array = input.split("\n");

    return input_array;
}


function part_1(): number {
    const data: string = read_file_lines()[0];
    let packet_start_index = 0;
    let start_index: number = 0;
    let end_index: number = 4;

    while (packet_start_index == 0) {
        let current_substring: string = data.substring(start_index, end_index);
        let current_substring_as_array: string[] = [...current_substring];
        let current_substring_as_set = new Set(current_substring_as_array);

        if (current_substring_as_array.length == Array.from(current_substring_as_set).length) {
            packet_start_index = end_index;
            break;
        } else {
            start_index += 1;
            end_index += 1;
        }
    }

    return packet_start_index;
}


function part_2(): number {
    const data: string = read_file_lines()[0];
    let message_start_index = 0;
    let start_index: number = 0;
    let end_index: number = 14;

    while (message_start_index == 0) {
        let current_substring: string = data.substring(start_index, end_index);
        let current_substring_as_array: string[] = [...current_substring];
        let current_substring_as_set = new Set(current_substring_as_array);

        if (current_substring_as_array.length == Array.from(current_substring_as_set).length) {
            message_start_index = end_index;
            break;
        } else {
            start_index += 1;
            end_index += 1;
        }
    }

    return message_start_index;
}


console.log(`Part 1: ${part_1()}`);
console.log(`Part 2: ${part_2()}`);

File: Advent of Code/2022/day 6/solution.dart
import 'dart:convert';
import 'dart:io';

Future<List<String>> read_input_file({String filename = "./input.txt"}) async {
  /*
    Read `filename` and return an array containing each line of the file. Each line
      in the file is returned as an entry in the array.
  */
  List<String> file_contents = [];
  final file = File(filename);

  if (await file.exists()) {
    final linesStream = file.openRead();
    await for (var line
        in linesStream.transform(Utf8Decoder()).transform(LineSplitter())) {
      file_contents.add(line);
    }
  }

  return file_contents;
}

void part_1() async {
  // Read the input file
  var file_contents = await read_input_file();
  String input_string = file_contents[0];

  // Define variables
  int packet_start_index = 0;
  int start_index = 0;
  int end_index = 4;

  while (packet_start_index == 0) {
    var current_substring = input_string.substring(start_index, end_index);
    List<String> current_substring_as_array = current_substring.split("");
    Set<String> current_substring_as_set = current_substring_as_array.toSet();

    if (current_substring_as_array.length == current_substring_as_set.length) {
      packet_start_index = end_index;
      break;
    } else {
      start_index += 1;
      end_index += 1;
    }
  }

  print(packet_start_index);
}

void part_2() async {
  // Read the input file
  var file_contents = await read_input_file();
  String input_string = file_contents[0];

  // Define variables
  int packet_start_index = 0;
  int start_index = 0;
  int end_index = 14;

  while (packet_start_index == 0) {
    var current_substring = input_string.substring(start_index, end_index);
    List<String> current_substring_as_array = current_substring.split("");
    Set<String> current_substring_as_set = current_substring_as_array.toSet();

    if (current_substring_as_array.length == current_substring_as_set.length) {
      packet_start_index = end_index;
      break;
    } else {
      start_index += 1;
      end_index += 1;
    }
  }

  print(packet_start_index);
}

void main() {
  part_1();
  part_2();
}
